### PR TITLE
V8: Fix listviews in pickers

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes: https://github.com/umbraco/Umbraco-CMS/issues/3768

### Description

The picker properties aren't too fond of listviews:

1. The right arrow for expanding doesn't show (but you can double click your way to the children).
2. The checkmarks for selected items do not show in a multi picker scenario.

For example, when a page is configured to show its children in a listview, the content picker looks like this:

image

The member picker suffers from the same issue - only worse, because *every* node is a listview:

listview-member-picker-before

*Note: The member picker explodes if you select anything but "All Members" due to a missing Examine index. Fixing that is not in scope for this PR.*

This PR fixes the two problems mentioned above, and as a bonus it also fixes 2nd level list views and beyond 😄 

### Testing this PR

1. Create a doctype with children shown in a listview.
2. Create another doctype with children shown in the tree underneath.
3. Create some content based on both doctypes (make sure to create some children to each doctype).
4. Using a Content Picker, verify that you can navigate the entire content tree and pick children of each doctype.
5. Using a Multinode Treepicker set to pick content:
   - Verify that you can navigate the entire content tree and pick multiple children of each doctype.
   - Verify that the selected items are marked with the green checkmark.
6. Using a Media Picker, verify that you can still pick media as usual - or in other words, verify that this PR doesn't break the Media Picker.
7. Create some members.
8. Using a Member Picker, verify that you can expand "All Members" by using the right arrow and that you can pick a member.

It all comes together something like this:

listview-picker-after

🎄 17/24 🎄

### CODE REVIEW

**Comment by nul800sebastiaan** [comment]
Awesome, also with the member fixes! 🙌🤘

File: src/Umbraco.Web/Trees/ContentTreeController.cs [hunk 1]
```diff
@@ -61,14 +61,12 @@ protected override TreeNode GetSingleTreeNode(IEntitySlim entity, string parentI
                 //Special check to see if it ia a container, if so then we'll hide children.
                 var isContainer = entity.IsContainer;   // && (queryStrings.Get("isDialog") != "true");
 
-                var hasChildren = ShouldRenderChildrenOfContainer(entity);
```
Comment (Shazwazza) [inline src/Umbraco.Web/Trees/ContentTreeController.cs]: @kjac  + @nul800sebastiaan 

I'm pretty apprehensive about removing this! This check performs quite a lot of checks and security checks and now we're just ignoring all of that. The last time this was changed almost all the tree security was broken. 

This needs to be tested and re-tested asap

File: src/Umbraco.Web.UI.Client/src/common/services/tree.service.js [hunk 2]
```diff
@@ -287,7 +287,10 @@ function treeService($q, treeResource, iconHelper, notificationsService, eventsS
                 throw "No node defined on args object for loadNodeChildren";
             }
 
-            this.removeChildNodes(args.node);
+            // don't remove the children for container nodes in dialogs, as it'll remove the right arrow indicator
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/common/services/tree.service.js]: Isn't this just a symptom of a different problem?

File: src/Umbraco.Web/Trees/ContentTreeController.cs (hunk 1 above)
Comment (kjac) [reply src/Umbraco.Web/Trees/ContentTreeController.cs]: `ShouldRenderChildrenOfContainer` is still invoked in `GetTreeNodesInternal` to ensure that users with custom start nodes don't get stranded. In its current implementation it never acts as a "limiter" for child nodes based on security - this is handled within `PerformGetTreeNodes`. It does however always return `false` for container nodes, unless the current user's start node is set underneath the container. 

As such, `ShouldRenderChildrenOfContainer` explicitly tells the clientside that container nodes never have children (unless the current user's start node is set underneath the container), based on the assumption that the client doesn't want to render the container children in the tree view (due to the potential volume). This assumption works for tree structures, because they don't need to hint that there are children to a container node. But the pickers do - they have to show the right arrow for both normal and container nodes. Therefore the assumption doesn't work.

What I have done is basically letting the clientside know if any given node has children without assuming anything about the application of this information. From there it's up to the client to render in the most performant way - i.e. never fetching all children for a container node.

Long story short: I can't really see a security issue here. That may be due to my lack of history on the project. If you can outline the scenario that causes a problem, of course I'll be happy to work it out 😄

File: src/Umbraco.Web.UI.Client/src/common/services/tree.service.js (hunk 2 above)
Comment (kjac) [reply src/Umbraco.Web.UI.Client/src/common/services/tree.service.js]: Well... it's a workaround for because `getChildren()` never returns any children for container nodes due to performance considerations. You can say it's a symptom or a derivative effect of mixing two navigation paradigms (tree and list view) in one navigation.

I would love to let `getChildren()` return something else for container nodes, but I'm afraid it'll break too many things at this point.
